lncore/addr.go

```diff
@@ -0,0 +1,319 @@
+package lncore
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"strconv"
+	"encoding/hex"
+
+	"github.com/mit-dci/lit/bech32"
+	"github.com/mit-dci/lit/crypto/fastsha256"
+)
+
+/*
+ * Ok so there's a lot of stuff going on here.  Lit natively supports addresses
+ * that are just the pkh of the pubkey encoded in bech32.  And that's pretty
+ * cool, you look up the IP+port in a tracker (or DHT) and connected with
+ * Noise_XX, then verify that the pubkey they provide you matches the pkh you
+ * already have.  Also you can provide IP+port if you want.
+ *
+ * But in mainstream BOLT nodes they use the full pubkey hex-encoded.  They have
+ * a DHT where you look up the IP+port (if not provided) and connect with
+ * Noise_XK, which assumes you already have the pubkey (since you do) and that's
+ * all fine.
+ *
+ * This tries to unify all of them with formats like this under 1 parser:
+ *
+ *      <bech32 pkh>[@<ip>[:<port>]
+ * or
+ *
+ *      <hex of bech32 pubkey>[@<ip>[:<port>]
+ *
+ * or fully
+ *
+ *      <bech32 pkh>:<pubkey>[@<ip>:<port>]
+ */
+
+// LnAddr is just a bech32-encoded pubkey.
+// TODO Move this to another package so it's more obviously not *just* IO-related.
+type LnAddr string
+
+// ToString returns the LnAddr as a string.  Right now it just unwraps it but it
+// might do something more eventually.
+func (lnaddr LnAddr) ToString() string {
+	return string(lnaddr)
+}
+
+// LnDefaultPort is from BOLT1.
+const LnDefaultPort = 9735
+
+// PkhBech32Prefix if for addresses lit already uses.
+const PkhBech32Prefix = "ln"
+
+// PubkeyBech32Prefix is for encoding the full pubkey in bech32.
+const PubkeyBech32Prefix = "lnpk"
+
+// LnAddressData is all of the data that can be encoded in a parsed address.
+type LnAddressData struct {
+	// Pkh is the pubkey hash as bech32.
+	Pkh    *string
+	
+	// Pubkey is the pubkey, as bytes, decoded from source format.
+	Pubkey []byte
+
+	// IPAddr is the IP address, if present.
+	IPAddr *string
+
+	// Port should always be defined, default is LnDefaultPort.
+	Port   uint16
+}
+
+// HasFullAddress returns whether or not we need to do a tracker/DHT lookup in
+// order to resolve the connection address for this address data.  Basically, it
+// checks if there's an IP address defined.
+func (d *LnAddressData) HasFullAddress() bool {
+	return d.IPAddr != nil
+}
+
+// AddrFmtFull is <bech32 pkh>:<hex pk>@<ip>:<port>
+const AddrFmtFull = "full"
+
+// AddrFmtFullBech32Pk is <bech32 pkh>:<bech32 pk>@<ip>:<port>
+const AddrFmtFullBech32Pk = "full_bech32_pk"
```

Comment:
```
// AddrFmtFullBech32Pk is <bech32 pkh>:<bech32 pk>@<ip>:<port>
```
in the above, if we have the bech32 pk, we could easily derive the bech32 kph (`LitAddr`). Maybe we could just have the bech32 pk in full and then use the relevant methods where we need the pkh?

Reply:
Oh yes you can definitely derive the pkh from the pk.  I hope nobody uses this format except when making fully-qualified addresses for whatever reason.  But in practical use I don't really see why anyone would use it.  I left it (as with the variant with the hex pk) just as "one option".